src/we_aes_gcm.c

```diff
@@ -356,153 +302,142 @@ static int we_aes_gcm_cipher(EVP_CIPHER_CTX *ctx, unsigned char *out,
-                if (aes->decryptBuf == NULL) {
-                    WOLFENGINE_ERROR_FUNC_NULL(WE_LOG_CIPHER,
-                        "OPENSSL_malloc", aes->decryptBuf);
+            if (out == NULL && len != 0 && in != NULL) {
+                WOLFENGINE_MSG(WE_LOG_CIPHER, "Resizing stored AAD and "
+                    "appending data, len = %d", (int)len);
+                /* Resize stored AAD and append new data. */
+                p = (unsigned char*)OPENSSL_realloc(aes->aad, aes->aadLen +
+                        (int)len);
+                if (p == NULL) {
+                    WOLFENGINE_ERROR_FUNC_NULL(WE_LOG_CIPHER, "OPENSSL_realloc",
+                        p);
                     ret = -1;
                 }
                 else {
-                    XMEMCPY(aes->decryptBuf, in, len);
-                    aes->decryptBufLen = len;
+                    /* Copy in new data after existing data. */
+                    aes->aad = p;
+                    XMEMCPY(aes->aad + aes->aadLen, in, len);
+                    aes->aadLen += len;
+                    ret = (int)len;
                 }
-                /* No data decrypted, yet, so return 0. */
-                ret = 0;
-            }
-            else
-        #endif
-            {
-                ret = we_aes_gcm_decrypt(aes, out, in, len);
             }
-        }
-
-    #ifdef WE_AES_GCM_DECRYPT_ON_FINAL
-        if (aes->enc || (!aes->enc && aes->tagLen > 0))
-    #endif
-        {
-            /* Dispose of any AAD - all used now. */
-            OPENSSL_free(aes->aad);
-            aes->aad = NULL;
-            aes->aadLen = 0;
-        }
-        if (ret == 1) {
-            if (aes->ivInc) {
-                int i;
-                for (i = aes->ivLen - 1; i >= aes->ivLen - 8; i--) {
-                    if ((++aes->iv[i]) != 0) {
-                        break;
+            else {
+                if (!aes->ivSet) {
+                    /* Set extern IV. */
+                    WOLFENGINE_MSG(WE_LOG_CIPHER, "Setting external IV");
+                    rc = wc_AesGcmSetExtIV(&aes->aes, aes->iv, aes->ivLen);
+                    if (rc != 0) {
+                        WOLFENGINE_ERROR_FUNC(WE_LOG_CIPHER,
+                            "wc_AesGcmSetExtIV", rc);
+                        ret = -1;
+                    }
+                    else {
+                        aes->ivSet = 1;
                     }
                 }
-                aes->ivInc = 0;
-                aes->ivSet = 0;
-            }
-            ret = (int)len;
-        }
-    }
-    else if ((ret == 1) && (len == 0)) {
-    #ifdef WE_AES_GCM_DECRYPT_ON_FINAL
-        if (!aes->enc) {
-            ret = we_aes_gcm_decrypt(aes, out, aes->decryptBuf,
-                aes->decryptBufLen);
-            aes->tagLen = 0;
-        }
-    #endif
+                if (ret == 1 && (len != 0 || in != NULL)) {
+                    /* "Update" called. */
```

Comment:
Does EVP_Cipher() work?

Reply:
Arg, I knew I missed something. I need to add a unit test that uses EVP_Cipher(). I'm not sure if this redesign is compatible with it. I don't think it is, so I may just have to revert to the last implementation and tweak it a bit.

Reply:
Actually, the old implementation with `WE_AES_GCM_DECRYPT_ON_FINAL` defined wouldn't work with EVP_Cipher() either, strictly speaking. It does work if you use it like OpenSSH does:

```
if (authlen) {
    u_char lastiv[1];

    if (authlen != cipher_authlen(cc->cipher))
        return SSH_ERR_INVALID_ARGUMENT;
    /* increment IV */
    if (!EVP_CIPHER_CTX_ctrl(cc->evp, EVP_CTRL_GCM_IV_GEN,
        1, lastiv))
        return SSH_ERR_LIBCRYPTO_ERROR;
    /* set tag on decyption */
    if (!cc->encrypt &&
        !EVP_CIPHER_CTX_ctrl(cc->evp, EVP_CTRL_GCM_SET_TAG,
        authlen, (u_char *)src + aadlen + len))
        return SSH_ERR_LIBCRYPTO_ERROR;
}
if (aadlen) {
    if (authlen &&
        EVP_Cipher(cc->evp, NULL, (u_char *)src, aadlen) < 0)
        return SSH_ERR_LIBCRYPTO_ERROR;
    memcpy(dest, src, aadlen);
}
if (len % cc->cipher->block_size)
    return SSH_ERR_INVALID_ARGUMENT;
if (EVP_Cipher(cc->evp, dest + aadlen, (u_char *)src + aadlen,
    len) < 0)
    return SSH_ERR_LIBCRYPTO_ERROR;
if (authlen) {
    /* compute tag (on encrypt) or verify tag (on decrypt) */
    if (EVP_Cipher(cc->evp, NULL, NULL, 0) < 0)
        return cc->encrypt ?
            SSH_ERR_LIBCRYPTO_ERROR : SSH_ERR_MAC_INVALID;
    if (cc->encrypt &&
        !EVP_CIPHER_CTX_ctrl(cc->evp, EVP_CTRL_GCM_GET_TAG,
        authlen, dest + aadlen + len))
        return SSH_ERR_LIBCRYPTO_ERROR;
}
```

Here, they validate the tag and effectively call "final" by calling EVP_Cipher() with NULL input and output. Apparently that's the way to validate the tag for AES-GCM with this interface. Note that this OpenSSH code works just fine with this new implementation, too.

OpenSSL's AES-GCM implementation will successfully decrypt ciphertext even if no tag's been set (i.e. even if the EVP_CTRL_GCM_SET_TAG command shown above was never issued). As we've discussed before, with wolfSSL there's no way to avoid the tag check (without using the newish AES-GCM streaming code, which FIPS v2 can't use). So, there's really no way that I can see for us to make EVP_Cipher() work properly in a single shot UNLESS the tag has been set. This all gets quite confusing, but in summary:

- The old implementation was subject to the same problems with the EVP_Cipher() interface.
- This implementation passes all the same tests locally on my machine (OpenSSL unit tests, wolfEngine unit tests, OpenSSH tests) AND it resolves this curl problem.

Let me know what you think @SparkiDev.

Reply:
OK,. So calling EVP_CIPHER is like calling update.
Add a test case into test_aestag.c that shows the call sequence that we support and I'll be happy.